Add a Helm chart

takeshi-cloudnatix opened the pull request.

takeshi-cloudnatix commented on deployments/server/templates/_helpers.tpl:
```diff
@@ -0,0 +1,63 @@
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "frontend-server.name" -}}
```
picked `frontend-server` than `frontend` for consistency with the other Helm charts. not a strong opinion

takeshi-cloudnatix commented on deployments/server/templates/deployment.yaml:
```diff
@@ -0,0 +1,48 @@
+          image: "{{ .Values.image.repository }}:{{ .Values.version }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          command: ["serve"]
+          args: ["-s", "build-{{ .Values.env }}", "-l", "80"]
```
for now, keeping the prod/staging/dev variation switching

takeshi-cloudnatix commented on deployments/server/templates/ingress.yaml:
```diff
@@ -0,0 +1,28 @@
+  labels:
+    {{- include "frontend-server.labels" . | nindent 4 }}
+  annotations:
+    {{- toYaml .Values.ingress.annotations | nindent 4 }}
```
moved to the values file

kkaneda replied:
We would like to use `global` so that the common values can be used across multiple charts. Could you update accordingly? Here is an example.

https://github.com/llmariner/inference-manager/blob/main/deployments/server/templates/ingress.yaml#L8

takeshi-cloudnatix replied:
Got it. Yeah, I'm aware of that but I wondered if it's ok to use the same configuration from `global` for the web UI, and kept them independent.

takeshi-cloudnatix commented on deployments/server/templates/ingress.yaml:
```diff
@@ -0,0 +1,28 @@
+    secretName: {{ . }}
+    {{- end }}
+  rules:
+  - host: {{ .Values.ingress.host }}
```
I guess just having `host` here works, but haven't checked yet

kkaneda replied:
This one is special and we don't want to use `global` as `global.host` is set to the API endpoint while this one is the frontend.

takeshi-cloudnatix replied:
Yes, that's why I kept this independent.

To make sure, what I meant by the initial comment was that I guess but not sure that we don't need to put the `host` field to the other services to explicitly instruct their Ingress to handle only `api.xxx` host, but just having `app.xxx` here would make things work given that they have the `path` field to capture only the `/v1/the_api` field.

takeshi-cloudnatix commented on deployments/server/values.yaml:
```diff
@@ -0,0 +1,141 @@
+# frontend-server is a dependency of LLMariner, so global
+# values are propagated at its installation.
+# Please see https://github.com/llmariner/llmariner/blob/main/deployments/llmariner.
+global:
```
Nothing is defined for now. To be used.

Reference: https://helm.sh/docs/chart_template_guide/subcharts_and_globals/#global-chart-values

takeshi-cloudnatix commented on deployments/server/templates/deployment.yaml:
```diff
@@ -0,0 +1,48 @@
+              containerPort: {{ .Values.httpPort }}
+              protocol: TCP
+          env:
+          - name: npm_package_version
```
Note: used by Sentry

https://github.com/cloudnatix/frontend/blob/e0db9fba0d8a9c20503e2c87a7becdce2f410e8c/src/next/utils/sentry.ts#L42

takeshi-cloudnatix commented on deployments/server/values.yaml:
Based on https://github.com/cloudnatix/frontend/blob/main/ci/webui/values.yaml, but formatted referring to Aya's work e.g. https://github.com/llmariner/vector-store-manager/blob/main/deployments/server/values.yaml.

takeshi-cloudnatix commented on deployments/server/values.yaml:
```diff
@@ -0,0 +1,141 @@
+  # The container image name.
+  repository: public.ecr.aws/cloudnatix/llmariner/frontend
+  # Kubernetes imagePullPolicy on Deployment.
+  pullPolicy: IfNotPresent
```
changed to `IfNotPresent` from `Always`

takeshi-cloudnatix commented:
Nothing has been checked with Saniya's WIP change.

I'd like to get a high level review now to see if I'm understanding the convention of and how things are working for LLMariner.

takeshi-cloudnatix commented:
Aya, please focus on general issues as a Helm chart.

Kenji, please focus on the details of the frontend.

takeshi-cloudnatix commented:
Schema generation is to do.

takeshi-cloudnatix commented:
CC: @saniya-cloudnatix . Just FYI. No need to review now.

kkaneda commented on deployments/server/templates/ingress.yaml:
```diff
@@ -0,0 +1,28 @@
+  annotations:
+    {{- toYaml .Values.ingress.annotations | nindent 4 }}
+spec:
+  ingressClassName: {{ .Values.ingress.ingressClassName }}
```
Probably better to use global here as well.

takeshi-cloudnatix replied:
done

kkaneda commented on deployments/server/values.yaml:
```diff
@@ -0,0 +1,141 @@
+
+serviceAccount:
+  # Specifies whether a service account should be created
+  create: true
```
We don't need a service account. I think we can remove the service account creation or set the default value to `false`.

takeshi-cloudnatix replied:
removed!

kkaneda commented on deployments/server/values.yaml:
```diff
@@ -0,0 +1,141 @@
+# The container image of frontend-server.
+image:
+  # The container image name.
+  repository: public.ecr.aws/cloudnatix/llmariner/frontend
```
Change this to `frontend-server` as well?

takeshi-cloudnatix replied:
done

kkaneda reviewed (commented):
Thanks!

takeshi-cloudnatix commented on deployments/server/README.md:
```diff
@@ -0,0 +1,42 @@
+
+## Configuration
+
+See [Customizing the Chart Before Installing](https://helm.sh/docs/intro/using_helm/#customizing-the-chart-before-installing). To see all configurable options with detailed comments, visit the chart's [values.yaml](./values.yaml), or run these configuration commands:
```
ah, if it's about the caption of the link to the Helm official documentation, I think it's intended to be aligned with what's there.

https://helm.sh/docs/intro/using_helm/#customizing-the-chart-before-installing

takeshi-cloudnatix commented on deployments/server/README.md:
````diff
@@ -0,0 +1,42 @@
+See [Customizing the Chart Before Installing](https://helm.sh/docs/intro/using_helm/#customizing-the-chart-before-installing). To see all configurable options with detailed comments, visit the chart's [values.yaml](./values.yaml), or run these configuration commands:
+
+```console
+helm show values oci://public.ecr.aws/cloudnatix/llmariner-charts/frontend
````
Thank you. changed

aya-cloudnatix commented on deployments/server/README.md:
```diff
@@ -0,0 +1,42 @@
+# Frontend
+
+[![Artifact Hub](https://img.shields.io/endpoint?url=https://artifacthub.io/badge/repository/frontend)](https://artifacthub.io/packages/search?repo=frontend)
```
Could you create an Artifact Hub account and give me your username? I'll add you to LLMariner.org, and then you can register a new repository and generate a badge tag for this.

takeshi-cloudnatix replied:
Created. It's `takeshi`. Thanks!

aya-cloudnatix replied:
Thanks! invited

takeshi-cloudnatix replied:
Done. Thanks!

aya-cloudnatix commented on deployments/server/values.yaml:
```diff
@@ -0,0 +1,141 @@
+# Default values for frontend-server.
+# This is a YAML-formatted file.
+# Declare variables to be passed into your templates.
+
```
```suggestion

# +docs:section=Global

```

takeshi-cloudnatix replied:
done!

aya-cloudnatix commented on deployments/server/templates/ingress.yaml:
```diff
@@ -0,0 +1,28 @@
+    {{- toYaml .Values.ingress.annotations | nindent 4 }}
+spec:
+  ingressClassName: {{ .Values.ingress.ingressClassName }}
+  {{- if .Values.ingress.tls }}
```
There seems to be no `end` for this block.
```
> helm template .
Error: parse error at (frontend-server/templates/ingress.yaml:29): unexpected EOF

Use --debug flag to render out invalid YAML
```

takeshi-cloudnatix replied:
fixed!

takeshi-cloudnatix reviewed (commented):
Thanks!

aya-cloudnatix reviewed (commented):
I think it would be good to add the linter.

takeshi-cloudnatix commented:
> I think it would be good to add the linter.

Just added and run the linter. Added `Makefile` rules for that. Fixed issues reported except for one.

takeshi-cloudnatix commented on deployments/server/values.yaml:
```diff
@@ -0,0 +1,134 @@
+  #   secretName: app-tls
+  #
+  # +docs:property
+  # tls:
```
@aya-cloudnatix 

Does the `+docs:property` annotation work also for the `lint` command?

```
make helm-lint
cd ./deployments/server && helm-tool lint
value missing from values.yaml: ingress.tls.secretName
value missing from templates: tls
Could not lint: values.yaml and templates are not in sync
make: *** [Makefile:21: helm-lint-server] Error 1
```

aya-cloudnatix replied:
It works, but the `+docs:property` value is parsed as a map. In this case, `secretName` is parsed as a map value. You can either uncomment it or insert it as a map without explicitly specifying the child elements in the template.

takeshi-cloudnatix replied:
> insert it as a map without explicitly specifying the child elements in the template.

Ah, yeah, this would work.

I did some more research to understand what's happening.

It looks like `+docs:property` works only when there's some uncommented field/entry after the `+docs:property`-annotated comment.

E.g. this works:

```
foo:
  # +docs:property
  # bar: 123

  baz: "abc"
```

but this doesn't work:

```
foo:
  # +docs:property
  # bar: 123

something:
  ...
```

The comment is handled as at the same level as `foo` and `something`. I confirmed this by `helm-tool render` result.

Though it's fragile, for now, I tried moving this above the required field `host`.

Now, `helm-tool` recognizes `tls` is recognized as a field under `ingress`.

I further want to get `secretName` but I just understood that it's out of the scope of `helm-tool`. `helm-tool` parses YAML-ish snippet and extract the (top-level) name of that. I was just misunderstanding this. So, `helm-tool` sees only `ingress.tls` and verifies that `ingress.tls` is used by the templates.

This is not something we want to do. We can uncomment the whole subtree from `tls` to make `secretName` also checked.

> You can either uncomment it

So, yeah, I'll do this.

kenji-cloudnatix reviewed (approved):
LG!
